core/macros/src/class.rs

```diff
@@ -0,0 +1,703 @@
+                    .ok_or( boa_engine::js_error!( #err ))?;
+            }
+        } else {
+            quote! {
+                let self_ = &*this.as_downcast_ref::< #class_ty >()
+                    .ok_or( boa_engine::js_error!( #err ))?;
+            }
+        };
+
+        Ok((downcast, quote! { self_ }))
+    }
+
+    /// Serializes an argument of form `pat: Type` into its declaration and call.
+    #[allow(clippy::unnecessary_wraps)]
+    fn arg_from_pat_type(
+        pat_type: &mut PatType,
+        i: usize,
+    ) -> SpannedResult<(TokenStream2, TokenStream2)> {
+        let ty = pat_type.ty.as_ref();
+        let ident = Ident::new(&format!("boa_arg_{i}"), Span::call_site());
+
+        // Find out if it's a boa context.
+        let is_context = match ty {
+            Type::Reference(syn::TypeReference {
+                elem,
+                mutability: Some(_),
+                ..
+            }) => match elem.as_ref() {
+                Type::Path(syn::TypePath { qself: _, path }) => {
+                    if let Some(maybe_ctx) = path.segments.last() {
+                        maybe_ctx.ident == "Context"
+                    } else {
+                        false
+                    }
+                }
+                _ => take_path_attr(&mut pat_type.attrs, "context"),
+            },
+            _ => false,
+        };
+
+        if is_context {
+            Ok((quote! {}, quote! { context }))
+        } else {
+            Ok((
+                quote! {
+                    let (#ident, rest): (#ty, &[boa_engine::JsValue]) =
+                        boa_engine::interop::TryFromJsArgument::try_from_js_argument( this, rest, context )?;
+                },
+                quote! { #ident },
+            ))
+        }
+    }
+
+    /// Create a `Function` from its name,
+    fn method(name: String, fn_: &mut ImplItemFn, class_ty: &Type) -> SpannedResult<Self> {
+        if fn_.sig.asyncness.is_some() {
+            error(&fn_.sig.asyncness, "Async methods are not supported.")?;
+        }
+
+        if !fn_.sig.generics.params.is_empty() {
+            error(&fn_.sig.generics, "Generic methods are not supported.")?;
+        }
+
+        let mut has_receiver = 0;
+        let (args_decl, args_call): (Vec<TokenStream2>, Vec<TokenStream2>) = fn_
+            .sig
+            .inputs
+            .iter_mut()
+            .enumerate()
+            .map(|(i, a)| match a {
+                FnArg::Receiver(receiver) => {
+                    has_receiver += 1;
+                    Self::arg_self_from_receiver(receiver, class_ty)
+                }
+                FnArg::Typed(ty) => Self::arg_from_pat_type(ty, i),
+            })
+            .collect::<SpannedResult<_>>()?;
+
+        let length =
+            take_length_from_attrs(&mut fn_.attrs)?.unwrap_or(args_decl.len() - has_receiver);
+        let fn_name = &fn_.sig.ident;
+
+        let is_static = take_path_attr(&mut fn_.attrs, "method") || has_receiver > 0;
```

Comment:
Isn't this suppose to be method, if it has a method attribute or it has self it should be a method, right?

Reply:
Yeah this was renamed to the inverse of what it was. Fixed.